### PR TITLE
Support on_failed callback for streaming rpc

### LINKED ISSUE #2285: stream感知正常和异常关闭
`StreamInputHandler`里面的`on_closed`在stream正常和异常关闭都会调用，希望加一个功能来感知是正常还是异常关闭。

### PR DESCRIPTION
**Pull request by chenBright**
### What problem does this PR solve?

Issue Number: resolve #2285 

Problem Summary:

### What is changed and the side effects?

Changed:

Side effects:
- Performance effects(性能影响):

- Breaking backward compatibility(向后兼容性): 

---
### Check List:
- Please make sure your changes are compilable(请确保你的更改可以通过编译).
- When providing us with a new feature, it is best to add related tests(如果你向我们增加一个新的功能, 请添加相关测试).
- Please follow [Contributor Covenant Code of Conduct](../../master/CODE_OF_CONDUCT.md).(请遵循贡献者准则).

### CODE REVIEW

File: src/brpc/stream.h [hunk 1]
```diff
@@ -44,7 +44,18 @@ class StreamInputHandler {
...
+    // 2. If `split_closed_and_failed` returns true,
+    //    2.1 `on_closed` will be called Whether `StreamClose`
+    //        is called by local side or remote side.
+    //    2.2 `on_failed` will be called when the stream is
+    //        closed abnormally.
+    virtual void on_closed(StreamId id) = 0;
+    virtual void on_failed(StreamId id, int error_code,
+                           const std::string& error_text);
+
+    virtual bool split_closed_and_failed() const { return false; }
```
Comment (wwbmmm) [inline src/brpc/stream.h:58]: 是不是可以不用这个参数，失败的时候会调用on_failed，然后调用on_closed

Comment (chenBright) [reply src/brpc/stream.h:58]: 可以，这样的流程更自然一些，使用也更方便。

Comment (chenBright) [reply src/brpc/stream.h]: done

Comment (chenBright) [reply src/brpc/stream.h]: done

**Comment by wwbmmm** [comment]
LGTM

**Pull request merged**